### PR TITLE
Feat: External Ci webhook new interface, Deployment separated from build node.

### PR DESCRIPTION
**Pull request by vikramdevtron**
# Description

External Ci with new interface, new interactions. flow medications on app workflow editor, deployment node (cd pipeline) is now can be created without build node (Ci pipeline) handled by backend and added provision for that.

- External CI (nodes) fetch api via App Id, list all the webhooks in app to draw workflow editor.
- External CI (node) fetch api via node id, it will return the webhook detail, url, payload etc. along with permission scope required for execution.
- Api token listing based on required permission scope, list all the tokens who have permission for that scope.
-  Cd pipeline create api modification, for external Ci api will create first WF and then create External Ci node and then create cd and link all of them in a WF.
- Cd pipeline delete will check wether parent node is external CI or not, it will delete external CI as well.
- External Ci fetch with the sample payload and permission scope.
- Api token listing based on scope provided, has to show all the tokens which are valid for trigger that webhook.
- Authenticator changes to understand the api-tokens from header.
- Fetch artifacts for deployment nodes in case of parent node is external Ci.

https://docs.google.com/document/d/1DXm-lNxhW9bVHqtWAqpJfo8BW3OEHMae6x4RczLUxe8/edit

Fixes : https://github.com/devtron-labs/devtron/issues/2665

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Refactoring 

# How Has This Been Tested?
- [x] Create external Ci webhook node with cd pipeline create api.
- [x] Attach deployment after deployment for external Ci pipeline.
- [x] Parallel deployment attached with external Ci webhook. 
- [x] With pre and post config deployment has been tested.
- [x] Cd pipeline deletes, if parent node is webhook, delete it as well, check weather any other parallel or nested pipeline exists or not.
- [x] App workflow should be deleted automatically if no nodes attached to WF.
- [x] Api token listing based on provided permission scope.
- [x] Webhook fetch api - sample payload and schema and description.
- [x] Fetch artifacts for deployment and parallel deployment when parent nodes is external ci
- [x] Fetch artifacts for pre and posd cd alongside deployment.
- [x] Auto Cd and Manual Cd Test case.
- [x] App workflow delete without deleting nodes.

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Review by kartik-579 — DISMISSED** [review]

**Review by kartik-579 — APPROVED** [review]

File: api/apiToken/ApiTokenRestHandler.go [hunk 1]
```diff
@@ -217,3 +218,36 @@ func (handler ApiTokenRestHandlerImpl) checkManagerAuth(token string, object str
 	}
 	return true
 }
+
+func (impl ApiTokenRestHandlerImpl) GetAllApiTokensForWebhook(w http.ResponseWriter, r *http.Request) {
```
Comment (pghildiyal) [inline api/apiToken/ApiTokenRestHandler.go:222]: why are we checking user token? it should be api token rt?

File: api/apiToken/ApiTokenRestHandler.go [hunk 2]
```diff
@@ -217,3 +218,36 @@ func (handler ApiTokenRestHandlerImpl) checkManagerAuth(token string, object str
...
+	appName := v.Get("appName")
+
+	// service call
+	res, err := impl.apiTokenService.GetAllApiTokensForWebhook(projectName, environmentName, appName, impl.CheckAuthorizationForWebhook)
+	if err != nil {
+		impl.logger.Errorw("service err, GetAllApiTokensForWebhook", "err", err)
+		common.WriteJsonResp(w, err, nil, http.StatusInternalServerError)
+		return
+	}
+	common.WriteJsonResp(w, err, res, http.StatusOK)
```
Comment (pghildiyal) [inline api/apiToken/ApiTokenRestHandler.go:242]: we are not checking authorization of api tokens?

File: api/restHandler/ExternalCiRestHandler.go [hunk 3]
```diff
@@ -89,3 +106,61 @@ func (impl ExternalCiRestHandlerImpl) HandleExternalCiWebhook(w http.ResponseWri
...
+	err = decoder.Decode(&req)
+	if err != nil {
+		impl.logger.Errorw("request err, HandleExternalCiWebhookByApiToken", "err", err, "payload", req)
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+	req.TriggeredBy = userId
+	impl.logger.Infow("request payload, HandleExternalCiWebhookByApiToken", "payload", req)
+
+	ciArtifactReq, err := impl.ciEventHandler.BuildCiArtifactRequestForWebhook(req)
```
Comment (pghildiyal) [inline api/restHandler/ExternalCiRestHandler.go:134]: too much logic in resthandler, it should be pushed down to service also name of functions need change

File: internal/sql/repository/CiArtifactRepository.go [hunk 4]
```diff
@@ -109,35 +114,54 @@ func (impl CiArtifactRepositoryImpl) GetByWfId(wfId int) (*CiArtifact, error) {
...
 	var artifactsB []Object
-	var queryTemp string = "SELECT cia.id, cia.material_info FROM ci_artifact cia" +
-		" INNER JOIN ci_pipeline cp on cp.id=cia.pipeline_id" +
-		" INNER JOIN pipeline p on p.ci_pipeline_id = cp.id" +
-		" WHERE p.id= ? ORDER BY cia.id DESC"
-	_, err = impl.dbConnection.Query(&artifactsB, queryTemp, cdPipelineId)
-	if err != nil {
-		return nil, err
+
+	if parentType == bean.WEBHOOK_WORKFLOW_TYPE {
```
Comment (pghildiyal) [inline internal/sql/repository/CiArtifactRepository.go:126]: too much logic in repository

Comment (vikramdevtron) [reply internal/sql/repository/CiArtifactRepository.go:126]: noted, we will re visit this section, it's a tested code so don't want much changes in this section, will do refactor this as a new activity.

File: api/restHandler/ExternalCiRestHandler.go (hunk 3 above)
Comment (vikramdevtron) [reply api/restHandler/ExternalCiRestHandler.go]: refactor done, only services calls and validation and auth in this rest handler.

File: api/apiToken/ApiTokenRestHandler.go (hunk 2 above)
Comment (vikramdevtron) [reply api/apiToken/ApiTokenRestHandler.go:242]: yes we are checking, passing auth function to the service.

**Review by kartik-579 — DISMISSED** [review]

**Review by kartik-579 — APPROVED** [review]

**Pull request merged**